extensions/s3/src/main/java/io/deephaven/extensions/s3/S3SeekableChannelProvider.java

```diff
@@ -130,8 +130,27 @@ public boolean isCompatibleWith(@NotNull final SeekableChannelContext channelCon
     }
 
     @Override
-    public SeekableByteChannel getWriteChannel(@NotNull final Path path, final boolean append) {
-        throw new UnsupportedOperationException("Writing to S3 is currently unsupported");
+    public SeekableByteChannel getWriteChannel(@NotNull final URI uri, final boolean append) {
+        throw new UnsupportedOperationException("Creating seekable write channels for S3 is currently unsupported, " +
+                "use getOutputStream instead");
+    }
+
+    @Override
+    public OutputStream getOutputStream(@NotNull final URI uri, final boolean append, final int bufferSizeHint) {
+        if (append) {
+            throw new UnsupportedOperationException("Appending to S3 is currently unsupported");
+        }
+        // bufferSizeHint is unused because s3 output stream is buffered internally into parts
+        return new S3OutputStream(uri, s3AsyncClient, s3Instructions);
+    }
+
+    @Override
+    public void abort(@NotNull final OutputStream outputStream) throws IOException {
```

Comment:
I guess given the interface here, I would propose a positive method `public void complete(OutputStream)` as opposed to `abort`, for reason mentioned earlier. A positive method could also work in the file case; we could imagine renaming from a temporary file into the final file name on complete.

Reply:
I wanted to propose the same thing. Basically, I want to move all the temporary file creation logic from ParquetTools to a new FileOutputStream which should make the ParquetTools code much cleaner.
But that would be a longer effort, so can be done independently from this issue.